fl4health/losses/mkmmd_loss.py

```diff
@@ -2,153 +2,202 @@
+        # h_u_componets should have the same shape
+        h_u_components = torch.exp((-1 * inner_product_quadruples) / (2 * torch.pow(gamma, 2)))
+        # Each column of h_u_components should now be u(x_{2i-1}, x_{2i}), u(y_{2i-1}, y_{2i}), u(x_{2i-1}, y_{2i}),
+        # and u(x_{2i}, y_{2i-1}), where u is the kernel_index^th kernel
+        # So we compute:
+        #   h_u[x_{2i-1}, x_{2i},y_{2i-1}, y_{2i}] =  u(x_{2i-1}, x_{2i}) + u(y_{2i-1}, y_{2i})
+        # - u(x_{2i-1}, y_{2i}) - u(x_{2i}, y_{2i-1})
+        h_u = h_u_components[:, 0] + h_u_components[:, 1] - h_u_components[:, 2] - h_u_components[:, 3]
+        # this results in a matrix of shape 1 x number of v_i_quadruples
+        return h_u.unsqueeze(0)
+
+    def compute_all_h_u_from_inner_products(self, inner_product_quadruples: torch.Tensor) -> torch.Tensor:
+        k_list = [
+            self.compute_h_u_from_inner_products(inner_product_quadruples, gamma.reshape(1)) for gamma in (self.gammas)
+        ]
+        # Matrix should be of shape number of kernels x number of quadruples, since we compute the kernel value on all
+        # quadrules for every kernel
+        return torch.cat(k_list)
 
-        n_k_vector: torch.Tensor = torch.zeros((1, self.kernel_num)).to(self.device)
-        batch_num = len(XX[0])
-
-        for i in range(0, batch_num - 1, 2):
-
-            n_k_vector += XX[:, i, i + 1] + YY[:, i, i + 1] - XY[:, i, i + 1] - YX[:, i, i + 1]
-
-        return 2 * n_k_vector / batch_num
-
-    def estimate_qk_vector(
-        self,
-        XX: torch.Tensor,
-        YY: torch.Tensor,
-        YX: torch.Tensor,
-        XY: torch.Tensor,
+    def compute_all_h_u_per_v_i(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
+        # v_i = [x_{2i-1}, x_{2i}, y_{2i-1}, y_{2i}]
+        v_is = self.construct_quadruples(X, Y)
+        # For the quadruples of the form (x,  x', y, y') we need distances for (x, x'), (y, y'), (x, y'), (x, y')
+        inner_product_quadruples = self.compute_euclidean_inner_products(v_is)
+        # all_h_u has shape number of kernels x number of quadruples
+        all_h_u = self.compute_all_h_u_from_inner_products(inner_product_quadruples)
+        return all_h_u
+
+    def compute_hat_d_per_kernel(self, all_h_u_per_v_i: torch.Tensor) -> torch.Tensor:
+        # all_h_u_per_vi has shape number kernels x number of v_is. Each row is h_u(v_i) for kernel u with the ith
+        # column being that kernel evaluated on v_i.
+        # output shape is number of kernels x 1
+        return torch.mean(all_h_u_per_v_i, dim=1, keepdim=True)
+
+    def compute_mkmmd(self, X: torch.Tensor, Y: torch.Tensor, beta: torch.Tensor) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features are the same size and that beta is a tensor of
+        # shape number of kernels x 1
+        all_h_u_per_vi = self.compute_all_h_u_per_v_i(X, Y)
+        hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_vi)
```

Comment:
Do we want to also compute estimated hat_d for mlmmd computation? In my original code I basically computed over all pairs and only used estimation (pairing i and i+1) for optimizing beta.
```
def compute_mmd(
        self,
        XX: torch.Tensor,
        YY: torch.Tensor,
        YX: torch.Tensor,
        XY: torch.Tensor,
    ) -> torch.Tensor:

        return XX.mean(dim=(1, 2)) + YY.mean(dim=(1, 2)) - XY.mean(dim=(1, 2)) - YX.mean(dim=(1, 2))
```

Reply:
hmmm...that's true about the single batch. I'd be good with forming XX, YY, XY for batch mkmmd calculations. I think that's a good suggestions.